client/src/main/java/io/pravega/client/netty/impl/SessionHandler.java

```diff
@@ -0,0 +1,208 @@
+
+    /**
+     * Close a session. This is invoked when the ClientConnection is closed.
+     * @param clientConnection Client Connection.
+     */
+    public void closeSession(ClientConnection clientConnection) {
+        int session = ((ClientConnectionImpl) clientConnection).getSession();
+        log.info("Closing Session: {} for Endpoint: {}", session, ((ClientConnectionImpl) clientConnection).getConnectionName());
+        sessionIdReplyProcessorMap.remove(session);
+    }
+
+    /**
+     * Set the Recent Message flag. This is used to avoid sending redundant KeepAlives over the connection.
+     */
+    void setRecentMessage() {
+        recentMessage.set(true);
+    }
+
+    /**
+     * Fetch the netty channel. If {@link Channel} is null then throw a ConnectionFailedException.
+     * @return  The current {@link Channel}
+     * @throws ConnectionFailedException Throw if connection is not established.
+     */
+    Channel getChannel() throws ConnectionFailedException {
+        Channel ch = channel.get();
+        if (ch == null) {
+            throw new ConnectionFailedException("Connection to " + connectionName + " is not established.");
+        }
+        return ch;
+    }
+
+    /**
+     * This function completes the input future when the channel is registered.
+     *
+     * @param future CompletableFuture which will be completed once the channel is registered.
+     */
+    void completeWhenRegistered(final CompletableFuture<Void> future) {
+        Preconditions.checkNotNull(future, "future");
+        registeredFutureLatch.register(future);
+    }
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+        super.channelRegistered(ctx);
+        Channel ch = ctx.channel();
+        channel.set(ch);
+        registeredFutureLatch.release(null); //release all futures waiting for channel registration to complete.
+        log.info("Connection established with Endpoint: {} on ChannelId: {}.", connectionName, ch);
+        ch.write(new WireCommands.Hello(WireCommands.WIRE_VERSION, WireCommands.OLDEST_COMPATIBLE_VERSION), ch.voidPromise());
+        // WireCommands.KeepAlive messages are sent for every network connection to a SegmentStore.
+        ScheduledFuture<?> old = keepAliveFuture.getAndSet(ch.eventLoop().scheduleWithFixedDelay(new KeepAliveTask(), 20, 10, TimeUnit.SECONDS));
+        if (old != null) {
+            old.cancel(false);
+        }
+    }
+
+    /**
+     * Invoke all the {@link ReplyProcessor#connectionDropped()} for all the registered sessions once the
+     * connection is disconnected.
+     *
+     * @see io.netty.channel.ChannelInboundHandler#channelUnregistered(ChannelHandlerContext)
+     */
+    @Override
+    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
+        registeredFutureLatch.reset();
+        ScheduledFuture<?> future = keepAliveFuture.get();
+        if (future != null) {
+            future.cancel(false);
```

Comment:
There are a lot of these concurrent members. It might be good if there were some docs at the top explaining in which order which sequences need to happen an why.